✨ Add Japanese Yen (JPY) currency support

mtane0412 opened the pull request:
This commit introduces support for the Japanese Yen (JPY) currency.  The following changes were made:

- Added JPY to the list of supported currencies in `apps/admin-x-settings/src/utils/currency.ts` and `ghost/admin/app/utils/currency.js`.
- Updated `currencyToDecimal` and `currencyFromDecimal` functions to handle zero-decimal currencies correctly.
- Updated `getNonDecimal` function to handle zero-decimal currencies correctly.
- Updated `getSubscriptionData` function to use the updated `getNonDecimal` function.
- Updated `PaymentsService` to handle zero-decimal currencies correctly when creating and retrieving prices from Stripe.
- Added JPY to the milestones test data in `ghost/core/test/unit/server/services/milestones/ MilestonesService.test.js`.

These changes ensure that the application correctly handles JPY and other zero-decimal currencies, improving accuracy and functionality.

Got some code for us? Awesome 🎊!

Please take a minute to explain the change you're making:
- Why are you making it?
  - I am running a Ghost media for Japanese and needed JPY as the currency, and since JPY is a Non-Decimal Currency on Stripe, I also needed to make sure that Stripe would process it correctly.
- What does it do?
  - Add JPY to Ghost's settlement currency and divide the amount by 100 to determine Non Decimal Currency and ensure that it is processed correctly on Stripe.
- Why is this something Ghost users or developers need?
  - There have been requests in past forums for JPY to be added, but it was not addressed for a long time. I also run a media outlet for Japanese people and needed the ability to settle payments in JPY.

Please check your PR against these items:

- [x] I've read and followed the [Contributor Guide](https://github.com/TryGhost/Ghost/blob/main/.github/CONTRIBUTING.md)
- [x] I've explained my change
- [x] I've written an automated test to prove my change works
  - I added the test but I am not confident. I need help.
  - 
We appreciate your contribution! 🙏

mtane0412 committed:
✨ Add Japanese Yen (JPY) currency support

This commit introduces support for the Japanese Yen (JPY)
currency.  The following changes were made:

- Added JPY to the list of supported currencies in
  `apps/admin-x-settings/src/utils/currency.ts` and
  `ghost/admin/app/utils/currency.js`.
- Updated `currencyToDecimal` and `currencyFromDecimal`
  functions to handle zero-decimal currencies correctly.
- Updated `getNonDecimal` function to handle zero-decimal
  currencies correctly.
- Updated `getSubscriptionData` function to use the
  updated `getNonDecimal` function.
- Updated `PaymentsService` to handle zero-decimal
  currencies correctly when creating and retrieving prices
  from Stripe.
- Added JPY to the milestones test data in
  `ghost/core/test/unit/server/services/milestones/
  MilestonesService.test.js`.

These changes ensure that the application correctly handles
JPY and other zero-decimal currencies, improving accuracy
and functionality.

mtane0412 committed:
✅  Add test case for adding initial JPY ARR milestone without sending email.

mtane0412 committed:
♻️ Refactor: Improve code readability in MilestonesService.test.js

The changes improve the readability of the code by adding line breaks to the asynchronous functions within the queries object.  This enhances maintainability and understanding of the test suite.

mtane0412 committed:
Merge branch 'main' into feature/add-jpy-currency

mtane0412 committed:
Merge branch 'main' into feature/add-jpy-currency

mtane0412 committed:
Merge branch 'main' into feature/add-jpy-currency

cathysarisky committed:
Merge branch 'main' into feature/add-jpy-currency

cathysarisky commented:
Hey @mtane0412 !

(I'm not a Ghost Foundation employee, just an open-source contributor. ) 

I can't tell you anything for sure, but I don't think this could be merged without a whole bunch of new tests.  (And indeed, you don't _want_ it to get merged without new tests, because you don't want to break something.)  For best success on getting the core team to evaluate and merge it, you should write a set of tests that demonstrate that it works.  I'd take a look at the existing tests, and basically try to add one that tests JPY in the same way other currencies are being tested.

Aside:  How's it look in Portal?  I'd be a little bit surprised if you didn't need some changes there, too...

mtane0412 committed:
Merge branch 'main' into feature/add-jpy-currency

mtane0412 committed:
✨ Add currency util tests: add unit tests for currency utils.

✅ Add tests for payments service with JPY currency: add tests to ensure correct handling of JPY currency in payments.

📝 Update context.json: update translation for free trial days.

📝 Update search.json: update Punjabi translation for search.

📝 Update signup-form.json: update Punjabi translation for signup form.

mtane0412 committed:
Merge branch 'main' into feature/add-jpy-currency

mtane0412 committed:
♻️ (admin): Refactor currency-test.js to improve readability and clarity of comments.

♻️ (core): Refactor payments.test.js to improve readability and clarity of comments.

mtane0412 commented:
@cathysarisky  Thanks so much for your help!

I’ve added the new tests and confirmed they all pass locally (http://localhost:4200/tests). The Stripe dashboard in my local environment now shows JPY amounts correctly.

Regarding Portal, no changes should be necessary there since this only affects the value sent to Stripe—the front-end display remains unchanged. I added the “divide by 100” step for two-decimal currencies on the back end so that USD is sent in cents (e.g. $100 → 10000) while JPY stays in whole yen (¥100 → 100).
If I'm wrong, please let me know.

cathysarisky committed:
Merge branch 'main' into feature/add-jpy-currency

cathysarisky commented:
Great! Thanks for adding tests!  Can you tidy up the whitespace changes on some of the i18n files?

mtane0412 committed:
Merge branch 'main' into feature/add-jpy-currency

mtane0412 committed:
Reverted: i18n files

mtane0412 commented:
@cathysarisky 
reverted.

mtane0412 committed:
Merge branch 'main' into feature/add-jpy-currency

cathysarisky commented:
Hey @mtane0412 , thanks for updating.

I think it would be wise to wait on some feedback from the Ghost core team (either here or on the forum) before you invest more time in this PR.  (Maybe I should have suggested that first before encouraging you to write tests, but it can be a bit of a chicken and egg problem.)

mtane0412 committed:
Merge branch 'main' into feature/add-jpy-currency

mtane0412 commented:
@cathysarisky 
I will do so. Thank you so much!

mtane0412 committed:
Merge branch 'main' into feature/add-jpy-currency

mtane0412 committed:
Merge remote-tracking branch 'upstream/main' into feature/add-jpy-currency

mtane0412 committed:
Merge remote-tracking branch 'upstream/main' into feature/add-jpy-currency

mtane0412 committed:
♻️ (PaymentsService.js): Improve handling of zero-decimal currencies in Stripe payments

The changes clarify the handling of zero-decimal currencies like JPY in Stripe payments.  The code now explicitly divides by 100 for these currencies to ensure compatibility with Stripe's expectations.  This improves the accuracy and reliability of payment processing.

mtane0412 committed:
🐛 Fixed import sorting in currency utils tests

Ensured imports are alphabetically ordered to satisfy ESLint’s ghost/sort-imports-es6 rule.

mtane0412 committed:
Merge branch 'main' into feature/add-jpy-currency

mtane0412 committed:
Merge branch 'main' into feature/add-jpy-currency

mtane0412 committed:
Merge branch 'main' into feature/add-jpy-currency

mtane0412 committed:
Merge branch 'main' into feature/add-jpy-currency

ErisDS commented:
Hey @mtane0412 I really appreciate the effort here, truly thank you. 

However, I encourage you to start on some smaller PRs first before trying to touch the payment system. As the most critical system in Ghost, which has businesses depending on it we wouldn't be able to accept changes from a first-time contributor.

I hope you understand 🙏 

I'm going to close this as I'm trying to get the list into a more manageable state, but if you're interested in getting to know the codebase a bit more, we can revisit in future.
